Potter/improve v2 connector docs (#3401)
Unstructured-IO/unstructured: unstructured/ingest/v2/README2.md

* context - The ProcessorConfig runs the pipeline. The arguments are related to the overall pipeline. We added some optional args to make development easier.
* source_connection - Takes arguments needed to connect to the source. Local files don't need anything here. Other connectors will.
* indexer - Takes the files in the `input_path` and creates .json files that point the downloader step to the right files 
* downloader - This does the actual downloading of the raw files (for non-blob files it may do something different like create a .txt file for every row in a source table)
* partitioner - Partitions the downloaded file provided it is a partionable file type.
* chunker/embedder - *Not represented here* but often needed to prepare files for upload to a vector database.
* stager - *Not represented here* but is often used to prepare partitioned files for upload.
* uploader - Uploads the blob-like files to the `output_dir`.

If you look at the folders/files in `local-working-dir` you will see the files that the pipeline creates as it runs.

```
local-working-dir
- index
  - a4a1035d57ed.json
- output
  - fake-text.txt.json
- partition
  - 36caa9b04378.json
```

> (Note that the index and partition file names are deterministic and based on the BLABLABLA) In the case of the local source connector, it won't *download* files because they are already local. But for other source connectors there will be a `download` folder. Also note that the final file is named based on the original file with a `.json` extension since it has been partitioned. Not all output files will be named the same as the input file. An example is a table as a source file, the output will be based on a hash of BLABLABLA.

Author: @roman can you leave a comment here on the two BLABLAs

Reviewer: The source connector chooses the filename of the downloaded content, but ever cached file along the way is a hash of the current step along with the previous hash via the filename.

You can see the source/destination connector file that it runs here:

https://github.com/Unstructured-IO/unstructured/blob/main/unstructured/ingest/v2/processes/connectors/local.py

> If you look through the file you will notice these Classes (actually @dataclasses because BLABLABLA) and functions

Author: @rbiseck3 can you leave a comment on the BLABLABLA

Reviewer: ```suggestion
If you look through the file you will notice these interfaces
```

* ChromaUploadStagerConfig - The Stager config. Didn't need anything for Chroma.

* ChromaUploadStager - The conform_dict is the critical method here. It takes the file we get from the Embedder step and prepares it for upload to the Chroma database. But it does not upload it. It saves the file to the `upload_stage` directory. The file type can be whatever makes sense for the Uploader phase.

* ChromaUploaderConfig - Attributes that are necessary for the upload stage specifically. The ChromaUploader will be upserting artifacts in batches.

* ChromaUploader - Connects to the Client. And uploads artifacts. Note that it does the minimum amount of processing possible to the artifacts before uploading. The Stager phase is responsible for preparing artifacts. Chroma wants artifacts in a dictionary of lists so we do have to create that in the Uploader since there is not a practical way to store that in a file.

* chroma_destination_entry - Registers the Chroma destination connector with the pipeline. (!!! LINK `unstructured/ingest/v2/processes/connectors/__init__.py`)

> Note that the `chroma.py` file imports the official Chroma python package when it *creates* the client and not at the top of the file. This is so that BLABLABLA

Author: @rbiseck3 can you leave a quick comment on the BLABLABLA

Reviewer: Note that the `chroma.py` file imports the official Chroma python package when it *creates* the client and not at the top of the file. This allows the classes to be instantiated without error and only cause a runtime error due to missing imports.

> # Developing V2 Connectors

Reviewer 2: Suggestion: Use "sentence case" for all headings instead of "Title Case."

## Intro

> The Unstructured open source repo processes documents (artifacts) in a pipeline. The Source and Destination connectors sit at the front and back of the pipeline. For a visual example see the flow diagram at the bottom (link to bottom).

Reviewer 2: Hmm...The "repo" doesn't process them. Perhaps "library" instead? Also, coming very soon, the open-source library won't be able to process them either, based on the move that was begun over to `unstructured-ingest`. Perhaps "Unstructured Ingest" instead?

Reviewer 2: Suggestion: Use lowercase for names of things, here and throughout, unless it is a product name or feature name that we typically use uppercase for. For instance, "source" and "destination" instead of "Source" and "Destination."

Reviewer 2: Don't forget the link here. Also, instead of "see the flow diagram at the bottom," provide the actual target heading name, to help readers more easily spot it at the end of this page.

> ## Simplest Example of a Pipeline

Reviewer 2: Minor: Perhaps "basic" instead of "simple?" Something that's simple to us might not be perceived as simple to others.

> The simplest example of a pipeline starts with a local source connector, followed by a partioner, and then ends with a local destination connector. Here is what the code to run this looks like:

Reviewer 2: "partioner" -> "partitioner"

>*** This is the type of Python file you'll want to run during development so that you can iterate on your connector.

`local.py`

```
from unstructured.ingest.v2.interfaces import ProcessorConfig
from unstructured.ingest.v2.pipeline.pipeline import Pipeline
from unstructured.ingest.v2.processes.connectors.local import (
    LocalConnectionConfig,
    LocalDownloaderConfig,
    LocalIndexerConfig,
    LocalUploaderConfig,
)
from unstructured.ingest.v2.processes.partitioner import PartitionerConfig

if __name__ == "__main__":
    Pipeline.from_configs(
        context=ProcessorConfig(
            verbose=True,
            work_dir="local-working-dir",
            reprocess=True,
            re_download=True,
        ),
        source_connection_config=LocalConnectionConfig(),
        indexer_config=LocalIndexerConfig(input_path="example-docs/fake-text.txt"),
        downloader_config=LocalDownloaderConfig(),
        partitioner_config=PartitionerConfig(),
        uploader_config=LocalUploaderConfig(output_dir="local-working-dir/output"),
    ).run()
```
You can run this with `python local.py` (Adjust the `input_path` and `output_dir` as appropriate.)

> The result is a partitioned `fake-text.txt.json` file in the `local-output` directory.

Reviewer 2: `fake-text.txt.json` or just `fake-test.json`?

Reviewer 2: I don't see a `local-output` directory in this code example?

Notice that the pipeline runs the following:

> * context - The ProcessorConfig runs the pipeline. The arguments are related to the overall pipeline. We added some optional args to make development easier.

Reviewer 2: Suggestion: Enclose the names of code symbols in backticks, here and throughout, for better readability.

Reviewer 2: > We added some optional args to make development easier.

Not sure what this means: which are optional, and why do they make development easier?

* source_connection - Takes arguments needed to connect to the source. Local files don't need anything here. Other connectors will.

> * indexer - Takes the files in the `input_path` and creates .json files that point the downloader step to the right files 

Reviewer 2: Minor: All sentences should end in a period, here and throughout.

* downloader - This does the actual downloading of the raw files (for non-blob files it may do something different like create a .txt file for every row in a source table)

> * partitioner - Partitions the downloaded file, provided it is a partionable file type. ([link to file types supported](https://github.com/Unstructured-IO/unstructured/blob/0c562d80503f6ef96504c6e38f27cfd9da8761df/unstructured/file_utils/filetype.py))

Reviewer 2: Minor: "link to file types supported" -> use a "see" or "view" phrase: i.e. "See the list of supported file types."

> * chunker/embedder - *Not represented here* but often needed to prepare files for upload to a vector database.

Reviewer 2: Minor: Should you add them to the code example, even if they're "empty" definitions?

> * stager - *Not represented here* but is often used to prepare partitioned files for upload.

Reviewer 2: Minor: Should you add it to the code example, even if it's an "empty" definition?

* uploader - Uploads the blob-like files to the `output_dir`.

If you look at the folders/files in `local-working-dir` you will see the files that the pipeline creates as it runs.

```
local-working-dir
- index
  - a4a1035d57ed.json
- output
  - fake-text.txt.json
- partition
  - 36caa9b04378.json
```

(Note that the index and partition file names are deterministic and based on the hash of the current step along with the previous step's hash.) In the case of the local source connector, it won't *download* files because they are already local. But for other source connectors there will be a `download` folder. Also note that the final file is named based on the original file with a `.json` extension since it has been partitioned. Not all output files will be named the same as the input file. This is the case for database like sources.

You can see the source/destination connector file that it runs here:

> https://github.com/Unstructured-IO/unstructured/blob/main/unstructured/ingest/v2/processes/connectors/local.py

Reviewer 2: Move over to `unstructured-ingest`?

If you look through the file you will notice these interfaces and functions

* LocalAccessConfig - This usually holds passwords, tokens, etc. This data gets hidden in all logs (and encrypted in our platform solution)

* LocalConnectionConfig - Username, host, port, etc. Anything needed for connecting to the service. It also imports the AccessConfig 

* LocalIndexerConfig - Holds attributes that allow Indexer to connect to the service and what kind of documents to filter for.

* LocalIndexer - Does the actual file listing and filtering. Note that it yields a FileData object.

* LocalDownloaderConfig - In this case it doesn't need anything if the LocalIndexerConfig already provides it.

* LocalDownloader - Does the actual downloading of the raw files.

* LocalUploaderConfig - Arguments for upload location

* LocalUploader - Does the actual uploading

* local_source_entry - Used to register the source connector here: `unstructured/ingest/v2/processes/connectors/__init__.py`

> * local_destination_entry - Used to register the destination connector here: `unstructured/ingest/v2/processes/connectors/__init__.py`

Reviewer 2: Move over to `unstructured-ingest`?
